src/tasks/buildAndUpload/index.ts

```diff
@@ -0,0 +1,80 @@
+import { ListrTask } from "listr/index.js";
+import { defaultVariantsDir } from "../../params.js";
+import { ListrContextBuildAndPublish } from "../../types.js";
+import { parseTimeout } from "../../utils/timeout.js";
+import {
+  getReleaseUploader,
+  cliArgsToReleaseUploaderProvider
+} from "../../releaseUploader/index.js";
+import { BuildAndUploadOptions } from "./types.js";
+import { buildVariantMap } from "./buildVariantMap.js";
+import { getVerifyConnectionTask } from "./getVerifyConnectionTask.js";
+import { getReleaseDirCreationTask } from "./getReleaseDirCreationTask.js";
+import { getFileValidationTask } from "./getFileValidationTask.js";
+import { getFileCopyTask } from "./getFileCopyTask.js";
+import { getBuildTasks } from "./getBuildTasks.js";
+import { getUploadTasks } from "./getUploadTasks.js";
+import { getDeleteOldPinsTask } from "./getDeleteOldPinsTask.js";
+import { getSaveUploadResultsTask } from "./getSaveUploadResultsTask.js";
+
+export function buildAndUpload({
+  contentProvider,
+  uploadTo,
+  userTimeout,
+  skipSave,
+  skipUpload,
+  requireGitData,
+  deleteOldPins,
+  composeFileName,
+  dir,
+  variantsDirPath = defaultVariantsDir,
+  variants
+}: BuildAndUploadOptions): ListrTask<ListrContextBuildAndPublish>[] {
+  const buildTimeout = parseTimeout(userTimeout);
+
+  // Release upload. Use function for return syntax
+  const releaseUploaderProvider = cliArgsToReleaseUploaderProvider({
+    uploadTo,
+    contentProvider
+  });
+  const releaseUploader = getReleaseUploader(releaseUploaderProvider);
+
+  const variantsMap = buildVariantMap({
+    variants,
+    rootDir: dir,
+    variantsDirPath,
+    composeFileName
+  });
+
+  return [
+    getFileValidationTask({ variantsMap, rootDir: dir }),
+    getVerifyConnectionTask({ releaseUploader, skipUpload }),
+    getReleaseDirCreationTask({ variantsMap }),
+    getFileCopyTask({
+      variantsMap,
+      rootDir: dir,
+      composeFileName,
+      requireGitData
+    }),
+    ...getBuildTasks({ variantsMap, buildTimeout, skipSave, rootDir: dir }),
+    ...getUploadTasks({
+      variantsMap,
+      releaseUploader,
+      requireGitData: !!requireGitData,
+      composeFileName,
+      skipUpload
+    }),
+    getDeleteOldPinsTask({
```

Comment:
does the "buildAndUpload" process stop if this `getDeleteOldPinsTask` fails? I would put it after `getSaveUploadResultsTask`, since it seems a cleanup task more than anything

Reply:
Actually, `getSaveUploadResultsTask` only stores the built hash to the `releases.json` file, so it is not critical either